mir_eval/display.py

```diff
@@ -0,0 +1,816 @@
+    ax.yaxis.set_minor_locator(MultipleLocator(1))
+
+    ax.axis('auto')
+    return ax
+
+
+def separation(sources, fs=22050, labels=None, ax=None, **kwargs):
+    '''Source-separation visualization
+
+    Parameters
+    ----------
+    sources : list of np.ndarray
+        A list of waveform buffers corresponding to each source
+
+    fs : number > 0
+        The sampling rate
+
+    labels : list of strings
+        An optional list of descriptors corresponding to each source
+
+    ax : matplotlib.pyplot.axes
+        An axis handle on which to draw the spectrograms.
+        If none is provided, a new set of axes is created.
+
+    kwargs :
+        Additional keyword arguments to `scipy.signal.spectrogram`
+
+    Returns
+    -------
+    ax
+        The axis handle for this plot
+    '''
+
+    # Get the axes handle
+    ax, _ = __get_axes(ax=ax)
+
+    if labels is None:
+        labels = ['Source {:d}'.format(_) for _ in range(len(sources))]
+
+    kwargs.setdefault('scaling', 'spectrum')
+
+    # The cumulative spectrogram across sources
+    # is used to establish the reference power
+    # for each individual source
+    cumspec = None
+    specs = []
+    for i, src in enumerate(sources):
+        freqs, times, spec = spectrogram(src, fs=fs, **kwargs)
+        specs.append(spec)
+        if cumspec is None:
+            cumspec = spec.copy()
+        else:
+            cumspec += spec
+
+    ref_max = cumspec.max()
+    ref_min = ref_max * 1e-6
+
+    legend_entries = []
+    for i, spec in enumerate(specs):
+
+        # For each source, grab a new color from the cycler
+        # Then construct a colormap that interpolates from
+        # [transparent white -> new color]
+        color = next(ax._get_lines.prop_cycler)['color']
+        cmap = LinearSegmentedColormap.from_list(labels[i],
+                                                 [(1.0, 1.0, 1.0, 0.0),
+                                                  color])
+
+        ax.pcolormesh(times, freqs, spec,
+                      cmap=cmap,
+                      norm=LogNorm(vmin=ref_min, vmax=ref_max),
+                      shading='gouraud',
+                      label=labels[i])
+
+        legend_entries.append(Patch(color=color, label=labels[i]))
+
+    handles, legend_labels = ax.get_legend_handles_labels()
+    handles.extend(legend_entries)
+    legend_labels.extend(labels)
+
+    ax.legend(handles=handles, labels=legend_labels)
+
+    __expand_limits(ax, [times.min(), times.max()], which='x')
+    __expand_limits(ax, [freqs.min(), freqs.max()], which='y')
+    return ax
+
+
+def __ticker_midi_note(x, pos):
+    '''A ticker function for midi notes.
+
+    Inputs x are interpreted as midi numbers, and converted
+    to [NOTE][OCTAVE]+[cents].
+    '''
+
+    NOTES = ['C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B']
+
+    idx = int(x % 12)
+    cents = float(np.mod(x, 1.0))
+    if cents >= 0.5:
+        cents = cents - 1.0
+        idx += 1
```

Comment:
> Related to this, maybe it would be nice if the lowest octave was configurable (I think -2 or -1 is more common than 0).

Really?  Isn't that well below the range of human pitch perception?